### PR TITLE
streams: fix pipe use after free

### PR DESCRIPTION
**Pull request by mikea**
found by new asan in upstream tests

### CODE REVIEW

**Comment by fhanau** [comment]
Glad you found this! Changes look good to me, but it would be helpful if someone with more async I/O experience can also take a look.

File: src/workerd/api/streams/internal.c++ [hunk 1]
```diff
@@ -1613,6 +1613,14 @@ bool WritableStreamInternalController::Pipe::checkSignal(jsg::Lock& js) {
   KJ_IF_SOME(signal, maybeSignal) {
     if ((signal)->getAborted()) {
       auto reason = signal->getReason(js);
+
+      // abort process might call parent.drain which will delete this,
+      // move/copy everything we need after into temps.
+      auto& parent = this->parent;
+      auto& source = this->source;
+      auto preventCancel = this->preventCancel;
+      auto promise = kj::mv(this->promise);
```
Comment (smerritt) [inline src/workerd/api/streams/internal.c++:1622]: Do we ever need to put `this->promise` back? We don't always call `parent.drain` here.

Comment (mikea) [reply src/workerd/api/streams/internal.c++:1622]: but we always call reject below.

Comment (mikea) [reply src/workerd/api/streams/internal.c++:1622]: same for source, we always reject it.

**Review by smerritt — APPROVED** [review]

**Pull request merged**